pkg/i2gw/providers/gce/gce_extensions.go

```diff
@@ -0,0 +1,188 @@
+		return "", false
+	}
+
+	return parseBackendConfigName(ctx, val)
+}
+
+// Get the backend config annotation from the K8s service if it exists.
+func getBackendConfigAnnotation(service *apiv1.Service, backendConfigKey string) (string, bool) {
+	val, ok := service.Annotations[backendConfigKey]
+	if ok {
+		return val, ok
+	}
+	return "", false
+}
+
+// Parse the name of the BackendConfig based on the annotation only if the same
+// BackendConfig is used for all Service ports.
+func parseBackendConfigName(ctx context.Context, val string) (string, bool) {
+	service := ctx.Value(serviceKey).(*apiv1.Service)
+
+	configs := annotations.BackendConfigs{}
+	if err := json.Unmarshal([]byte(val), &configs); err != nil {
+		notify(notifications.ErrorNotification, "BackendConfig annotation is invalid json", service)
+		return "", false
+	}
+
+	if configs.Default == "" && len(configs.Ports) == 0 {
+		notify(notifications.ErrorNotification, "No BackendConfig's found in annotation", service)
+		return "", false
+	}
+
+	if len(configs.Ports) != 0 {
+		notify(notifications.ErrorNotification, "Only config with default is supported since HealthCheckPolicy is attached on the whole service", service)
+		return "", false
+	}
+	return configs.Default, true
+}
+
+func buildGceServiceIR(ctx context.Context, storage *storage, ir *i2gw.IR) {
+	if ir.Services == nil {
+		ir.Services = make(map[types.NamespacedName]*i2gw.ServiceIR)
+	}
+
+	beConfigToSvcs := getBackendConfigMapping(ctx, storage)
+	for beConfigKey, beConfig := range storage.BackendConfigs {
+		if beConfig == nil {
+			continue
+		}
+		gceServiceIR := beConfigToGceServiceIR(beConfig)
+		serviceList := beConfigToSvcs[beConfigKey]
+		for _, svcKey := range serviceList {
+			if ir.Services[svcKey] == nil {
+				ir.Services[svcKey] = &i2gw.ServiceIR{}
+			}
+			ir.Services[svcKey].Gce = &gceServiceIR
+		}
+	}
+}
+
+func beConfigToGceServiceIR(beConfig *backendconfigv1.BackendConfig) ir.GceServiceIR {
+	var gceServiceIR ir.GceServiceIR
+	if beConfig.Spec.SessionAffinity != nil {
+		saConfig := ir.SessionAffinityConfig{
+			AffinityType: beConfig.Spec.SessionAffinity.AffinityType,
+			CookieTTLSec: beConfig.Spec.SessionAffinity.AffinityCookieTtlSec,
+		}
+		gceServiceIR.SessionAffinity = &saConfig
+	}
+
+	return gceServiceIR
+}
+
+func buildGceServiceExtensions(ir i2gw.IR, gatewayResources *i2gw.GatewayResources) {
+	for svcKey, serviceIR := range ir.Services {
+		if serviceIR == nil {
+			continue
+		}
+		bePolicy := toGCPBackendPolicy(svcKey, serviceIR)
+		if bePolicy == nil {
+			continue
+		}
+		obj, err := i2gw.CastToUnstructured(bePolicy)
+		if err != nil {
+			notify(notifications.ErrorNotification, "Failed to cast GCPBackendPolicy to unstructured", bePolicy)
+			continue
+		}
+		gatewayResources.GatewayExtensions = append(gatewayResources.GatewayExtensions, *obj)
+	}
+}
+
+func toGCPBackendPolicy(serviceNamespacedName types.NamespacedName, serviceIR *i2gw.ServiceIR) *gkegatewayv1.GCPBackendPolicy {
```

Comment:
```suggestion
func addBackendPolicyIfConfigured(serviceNamespacedName types.NamespacedName, serviceIR *i2gw.ServiceIR) *gkegatewayv1.GCPBackendPolicy {
```

Tiny nit: When I saw the `return nil` a couple lines below this if inputs were empty, I assumed that was an error case that we should be logging. Renaming the function to make it clear that this should only add a backendPolicy if it's specified, could help avoid confusion.

Reply:
Updated, thanks!